contractsV5/FeeHandler.sol

```diff
@@ -0,0 +1,102 @@
+        return (((((_burn << BITS_PER_PARAM) + _reward) << BITS_PER_PARAM) + _epoch) << BITS_PER_PARAM) + _expiryBlock;
+    }
+
+    function decodeData() public view returns(uint, uint, uint, uint) {
+        uint expiryBlock = brrAndEpochData & (1 << BITS_PER_PARAM) - 1;
+        uint epoch = (brrAndEpochData / (1 << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint rewardInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint burnInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        return (burnInBPS, rewardInBPS, epoch, expiryBlock);
+    }
+
+    // Todo: future optimisation to accumulate rebates for 2 rebate wallet
+    // encode totals, 128 bits per reward / rebate
+    function handleFees(address[] calldata eligibleWallets, uint[] calldata rebatePercentages) external payable returns(bool) {
+        // Decoding BRR data
+        (uint burnInBPS, uint rewardInBPS, uint epoch, uint expiryBlock) = decodeData();
+        uint rebateInBPS = BPS - rewardInBPS - burnInBPS;
+
+        // Check current block number
+        if(block.number > expiryBlock) {
+            (burnInBPS, rewardInBPS, rebateInBPS, epoch, expiryBlock) = kyberDAOContract.getLatestBRRData();
+
+            // Update brrAndEpochData
+            brrAndEpochData = encodeData(burnInBPS, rewardInBPS, epoch, expiryBlock);
+        }
+
+        uint rebateWei = rebateInBPS * msg.value / BPS;
+        uint rewardWei = rewardInBPS * msg.value / BPS;
+        for(uint i = 0; i < eligibleWallets.length; i ++) {
+            // Internal accounting for rebates per reserve wallet (totalRebatesPerRebateWallet)
+            totalRebatesPerRebateWallet[eligibleWallets[i]] = rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for total rebates (totalRebates)
+            totalRebates += rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for reward per epoch (totalRewardsPerEpoch)
+            totalRewardsPerEpoch[epoch] += rewardWei;
+            // Internal accounting for total rewards (totalRewards)
+            totalRewards += rewardWei;
+        }
+        // Todo: emit event
+        return true;
+    }
+
+
+    function claimStakerReward(address staker, uint percentageinPrecision, uint epoch) public {
+        // onlyDAO
+        // send reward
+        // update rewardPerEpoch
+        // update totalReward
+    }
+
+    function claimReserveRebate(address reserve) public {
+        // only DAO
+        // send rebate to reserve
+        // update rebatePerReserve;
+        // update total rebate amounts?
+        // update reserve rebate to 1 (avoid 0...) otherwise div by 0 issue? but will we even need to div by 0?
+        // if we include a dest address, we need an owner / admin of the reserve and the below function.
+    }
+
+    function burnKNC() public {
+        // only DAO?
```

Comment:
anyone can burn.

Reply:
yes anyone can call the burnKNC function from the DAO but the burnKNC function from the FeeHandler can only be called by the DAO

Reply:
I suppose it doesn't hurt to remove the modifier.

Reply:
I think the burning isn't related to the DAO
DAO only sets BRR values.
but burning is done by anyone.

we will have to limit amounts. per burn.
and create some block delay between burns.

Reply:
noted